### DISCUSSION TITLE
Leaking hostname and keys in filenames at VirusTotal

### DISCUSSION THREAD

**Discussion by danielmarschall** [rustdesk/rustdesk#5938] [General]
_(Posted to public after permission from RustDesk)_

I have found an issue in RustDesk that possibly affects security or privacy.

The feature "_put host and key into filename_" serves as a security risk, because filenames can be logged or obtained in various ways. For example

- In Taskmanager on multi-sser Systems. In Linux commands like `ps` or `w`
- History in the browser
- Local search indexes like `locate` on Linux
- Log or quarantine files of AntiVirus programs
- "Other file names" at VirusTotal.com

My greatest concern is the latter. When the program is uploaded to VirusTotal (**which can happen automatically if you use tools like Process Explorer**), VirusTotal will remember all the original filenames (containing the hosts and keys) and make them public. It will even be indexed by Google.

### Proof of concept:

1. Download RustDesk, e.g. version 1.2.0

2. Upload to VirusTotal

3. Click on "Details" and look at the original filenames.

Just for the Windows x64 build version 1.2.0 (`rustdesk-1.2.0-x86_64.exe`), 7 keys have been leaked through VirusTotal (also publicly indexed by Google):

| Host | Key |
|----------------------------|-------------------------|
| rustdesk.cloud.asmi.de     | qkdYZyBlZkFHTflQLQ2b... |
| rust.edv-marburg.de        | gnCZOUZF3wa04fB4p0+b... |
| remotedesktop.xgestevo.net | 254h6ZrSgAtoTHRFG+Ud... |
| rdesk.chert-tm.ru          | oD0JG6oEH6Vj52xxLm4H... |
| qs.isd.com.al              | tNiUZ+VbjpBhgWmqVDiE... |
| hbbs.bonzay.app            | 9HLdWlD+E4VunR9C0mG7... |
| rustdesk.sidamos.com       | lW2Mas77XtcO9+dKwmh...  |

(I have truncated the keys to avoid leaking them even further)

### Annotation:

To be fair, some people (probably unwittingly) leak their keys themselves, by putting Rustdesk with the host/key filename to their website. But I can understand that, because a lot of people who need help with the computer don't know how to use the command line, copy stuff from the clipboard, or unpack a ZIP file. So it is indeed very convenient that the host/key automatically gets applied by putting it in the filename.

On the other hand, you are basically making your server a "free" public server that everyone can use, maybe even for illegal activities like scamming people. In some countries, you get punished if you "help" people to commit a crime. In Germany, there is the "Störerhaftung" which means: If a scammer uses YOUR relay server to commit a crime, then YOU are partially liable for the damage since you "helped" commit the crime. There are exceptions for "safe harbor" / ISPs, but courts often have ruled against them, and a lot of people got their servers and computers taken away by police if they ran a TOR exit node that accessed prohibited Internet pages.

### Solutions which are NOT GOOD:

- As I wrote in the annotation, a lot of end users who need remote help do not know how to use things like the command line interface, therefore setting the parameters with the command line is not an option.

- Putting the parameters in a separate file (INI, TOML, BAT, PS1, ...) is also not a solution, because then you need to use a ZIP so you can download both. But then people might just unzip/execute the EXE itself (in Windows: Click the EXE inside the ZIP without unzipping all the contents first), or don't have a compression tool on their OS (if non-Windows).

- I also had the idea that the host/key is appended to the executable. In Windows, putting contents at the end of an EXE is legal. I am not sure about ELF files et. al.  However, then the AuthentiCode code signature becomes invalid, which is not good.

### Solutions which are GOOD:

I have no idea :-(

**Comment by rustdesk** [comment]
Thanks

**Comment by danielmarschall** [comment]
Wow, that was a quick response :) Sorry, I didn't know there was a Discussion section here at GitHub.


Two aspects that I have forgotten:

- One **_working_** solution would be: Adding information to the Executable through appending binary data (legal at least in Windows), PE section (AnyDesk does this), or Resource (Windows). **_And after that_**, the Executable needs to be automatically signed via a web service (AnyDesk and TeamViewer do this, but I understand this is a complicated task and I can understand that this would be a paid service only).

- The "pricing" page at rustdesk.com mentions "_Custom client generator (Coming soon)_" for the business self-hosting plan, which sounds like the solution for the host/key filename thing. HOWEVER, I think distributing RustDisk with a host/key pair is a core functionality and super useful and it would be a pity if only paid users could use this feature. Of course, you decide which things are free and which things are paid. In any case, your product is amazing!!

**Comment by danielmarschall** [comment]
FYI: I have sent a message to VirusTotal (Ticket 233066) notifying them that their web service is leaking hostnames and keys (linking to this GitHub discussion for more information), and asked if they could redact RustDesk keys from the publicly visible file names (e.g. replacing parts of it with underscores or something else). Once I receive a reply, I will report it here.

The relevant part in the email:

> My request [to VirusTotal] is not only to redact filenames of previous uploads but also to redact filenames of future uploads, e.g. by applying a RegEx in the VirusTotal GUI.
> 
> If a filename has the following format:  AAA-host=BBB-key=CCC.exe
> (where AAA is anything, BBB is the hostname and CCC is the base64 key)
> then please display the filename as: AAA-host=XYZ-key=XYZ.exe
> (where XYZ is anything that [...] is appropriate)
>

**Reply by rustdesk** [reply]
THANKS

**Comment by luckman212** [comment]
Any new advice? I just uploaded v1.2.3 to VT and it came back with a bunch of new hosts/keys... seems they haven't taken your advice yet to redact this info!

What can be done (maliciously) with the info here. Is it a big risk?

**Reply by danielmarschall** [reply]
Unfortunately, I did not get any reply from VirusTotal yet. Maybe I (or we) should write them more messages until they get annoyed? :-( 

Basically, with the keys people can use the server as relay. When "half of the world" is using your private server, you will notice that the performance is not as good as it should be with having an own server. In my opinion, the worst thing is that the relay server can be used to commit crimes (scamming, hacking, etc.), and depending on the local law, the server operator will either partially or fully punished, as if they had commited the crime themselves.

**Reply by danielmarschall** [reply]
I have contacted VirusTotal again. The new ticket number is 239418.

**Reply by luckman212** [reply]
Thanks @danielmarschall 

I reported it as well, Ticket# 239450

**Reply by luckman212** [reply]
Right - good point - done

**Comment by luckman212** [comment]
@danielmarschall FYI, I just got a reply from VirusTotal and they said they took care of this. Haven't had time to check & confirm yet!

**Reply by danielmarschall** [reply]
@luckman212 Sounds great!

I can verify that it works.

Version 1.2.3 now [shows up as](https://www.virustotal.com/gui/file/23b661d7bc171cd500d5096456905283ffe06479582b62d3bd5066633935d43e/details):
```
rustdesk-1.2.3-x86_64.exe
rustdesk
rustdesk.exe
RustDeskQS.exe
rustdesk-1.2.3-x86_64 (1).exe
303627ae-dcf1-4c97-83a5-f341c96c2e44
rustdesk-1.2.3-x86_64(1).exe
remoto.exe
rustdesk-1.2.3-x86_64(2).exe
rustdesk-1.2.3-x86_64(4).exe
rustdesk-1.2.3-x86_64(3).exe
rustdesk-licensed.exe
rustdesk-licensed(1).exe
rustdesk-host=rust.foto-it.com,key=.exe
rustdesk-host=rm.latinsoft.lv,key=32.exe
rustdesk-host=91.227.139.143,key=.exe
rustdesk-host=app.support.itfinity.de,key=-qs.exe
rustdesk-1.2.3-x86_64 (2).exe
rustdesk-host=rustdesk.mierisch-edv.com,key=.exe
RustDesk.exe
rustdesk_64.exe
rustdesk-host.exe
rustdesk-host=rustdesk.a2it.hr,key=.exe
rustdesk-1.2.3-x86_64 (3).exe
rustdesk-host=remote.devrdx.com,key=.exe
rustdesk-host=lukasz.ip-tech.net.pl,key=.exe
rustdesk-1.2.3-host=deskforge.de,key=.exe
rustdesk-host=rustdesk.dinabox.net,key=.exe
rustdesk-host=escritorio.dinabox.net,key=.exe
rustdesk-host=rd.lightserver.at,key=.exe
rustdesk-host=interlacerulez.no-ip.org,key=.exe
rustdesk-host=daddeldu,key=raffman,.exe
rustdesk-x86_64-host=meet.muon.de,key=.exe
1.EXE
rustdesk-host=rd.ussc.ru,key=.exe
RUSTDESK-host=rustdesk.flexiauction.com,key=,.EXE
rustdesk-1.2.3-x86_64--.exe
rustdesk64-1.2.3.exe
0b5a5130637d9a4fab2de35243416253.tmp
rustdesk-host=192.168.10.180,key==.exe
RustDesk 1.2.3.exe
rustdesk-1.TZW33l_V.2.3-x86_64.exe
```
The hostnames are still visible, but I think that's not a big deal.

---

Are newer versions also fixed? Yes.
I uploaded the new nightly 1.2.4 (it was not uploaded before) with a confidential filename.
[VT shows](https://www.virustotal.com/gui/file/9794726a1181c266e95573096e5b08e0b56d3fa6c138fcbe0c858b09552cec75/details) the following filenames (which is a bit weird, because I uploaded only one file, and I was the first person who uploaded it):
```
rustdesk-host
rustdesk
rustdesk.exe
```
